src/clean.c

```diff
@@ -0,0 +1,397 @@
+				return false;
+			}
+			break;
+		case 0:
+			/* Handle options that don't have shortcut. */
+			break;
+		default:
+			fprintf(stderr, "Error: unrecognized option: -%c,\n\n", opt);
+			return false;
+		}
+	}
+	return true;
+}
+
+typedef bool(remove_predicate_func)(const char *dir, const struct dirent *entry);
+
+/* Remove files from path for which pred returns true.
+ * Currently it doesn't recursively remove directories.
+ *
+ * Unless --all is set, the files are also filtered by their age.
+ */
+static int remove_if(const char *path, remove_predicate_func pred)
+{
+	int ret = 0;
+	DIR *dir;
+	const int seconds = DAYS_TO_KEEP_FILES * 60 * 60 * 24;
+
+	dir = opendir(path);
+	if (!dir) {
+		ret = errno;
+		goto end;
+	}
+
+	while (true) {
+		/* Reset errno to distinguish between a previous
+		 * failure and the end of stream. */
+		errno = 0;
+		struct dirent *entry;
+		entry = readdir(dir);
+		if (!entry) {
+			if (errno) {
+				ret = errno;
+			}
+			break;
+		}
+
+		if (!strcmp(entry->d_name, ".") || !strcmp(entry->d_name, "..")) {
+			continue;
+		}
+
+		char *file = NULL;
+		string_or_die(&file, "%s/%s", path, entry->d_name);
+
+		struct stat stat;
+		ret = lstat(file, &stat);
+		if (ret != 0) {
+			fprintf(stderr, "couldn't remove file %s: %s\n", file, strerror(errno));
+			goto next;
+		}
+
+		if (!options.all) {
+			if (now.tv_sec - stat.st_mtime <= seconds) {
+				goto next;
+			}
+		}
+
+		if (!pred(path, entry)) {
+			goto next;
+		}
+
+		if (options.dry_run) {
+			printf("%s\n", file);
+		} else {
+			if (S_ISDIR(stat.st_mode)) {
+				ret = rmdir(file);
+			} else {
+				ret = unlink(file);
+			}
+			if (ret != 0) {
+				fprintf(stderr, "couldn't remove file %s: %s\n", file, strerror(errno));
+			}
+		}
+		if (ret == 0) {
+			stats.files_removed++;
+		}
+	next:
+		free_string(&file);
+		ret = 0;
+	}
+
+end:
+	closedir(dir);
+	return ret;
+}
+
+static bool is_fullfile(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	return strlen(entry->d_name) == (SWUPD_HASH_LEN - 1);
+}
+
+static bool is_pack_indicator(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	static const char prefix[] = "pack-";
+	static const char suffix[] = ".tar";
+	static const size_t prefix_len = sizeof(prefix) - 1;
+	static const size_t suffix_len = sizeof(suffix) - 1;
+
+	const char *name = entry->d_name;
+	size_t len = strlen(name);
+	if (len < (prefix_len + suffix_len)) {
+		return false;
+	}
+	return !memcmp(name, prefix, prefix_len) && !memcmp(name + len - suffix_len, suffix, suffix_len);
+}
+
+static bool is_all_digits(const char *s)
+{
+	while (*s) {
+		if (!isdigit(*s)) {
+			return false;
+		}
+		s++;
+	}
+	return true;
+}
+
+static bool is_manifest(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	static const char prefix[] = "Manifest.";
+	static const size_t prefix_len = sizeof(prefix) - 1;
+
+	const char *name = entry->d_name;
+	if (strlen(name) < prefix_len) {
```

Comment:
I would have used strncmp rather than the 2 separate tests and to show that we are working on string data.

You only need the strlen(name) < prefix_len test to make sure you are not reading beyond the end of the name array,

Reply:
That is much better! Done.